src/common/internal/test_group.ts

```diff
@@ -446,40 +434,66 @@ class RunCaseSpecific implements RunCase {
     };
 
-    rec.start();
-    if (this.subcases) {
-      let totalCount = 0;
-      let skipCount = 0;
-      for (const subParams of this.subcases) {
-        rec.info(new Error('subcase: ' + stringifyPublicParams(subParams)));
-        try {
-          const params = mergeParams(this.params, subParams);
-          const subcaseQuery = new TestQuerySingleCase(
-            selfQuery.suite,
-            selfQuery.filePathParts,
-            selfQuery.testPathParts,
-            params
-          );
-          await this.runTest(rec, params, true, getExpectedStatus(subcaseQuery));
-        } catch (ex) {
-          if (ex instanceof SkipTestCase) {
-            // Convert SkipTestCase to info messages
-            ex.message = 'subcase skipped: ' + ex.message;
-            rec.info(ex);
-            ++skipCount;
-          } else {
-            // Since we are catching all error inside runTest(), this should never happen
-            rec.threw(ex);
+    try {
+      rec.start();
+      const inst = new this.fixture(rec, this.params);
+      try {
+        await inst.doInit();
+        if (this.beforeFn) {
+          await this.beforeFn(inst as Fixture & { params: {} });
+        }
+
+        if (this.subcases) {
+          let totalCount = 0;
+          let skipCount = 0;
+          for (const subParams of this.subcases) {
+            rec.info(new Error('subcase: ' + stringifyPublicParams(subParams)));
+            try {
+              const params = mergeParams(this.params, subParams);
+              const subcaseQuery = new TestQuerySingleCase(
+                selfQuery.suite,
+                selfQuery.filePathParts,
+                selfQuery.testPathParts,
+                params
+              );
+              inst.setParams(params);
+              await this.runTest(rec, inst, /* throwSkip */ true, getExpectedStatus(subcaseQuery));
+            } catch (ex) {
+              if (ex instanceof SkipTestCase) {
+                // Convert SkipTestCase to info messages
+                ex.message = 'subcase skipped: ' + ex.message;
+                rec.info(ex);
+                ++skipCount;
+              } else {
+                // Since we are catching all error inside runTest(), this should never happen
+                rec.threw(ex);
+              }
+            }
+            ++totalCount;
           }
+          if (skipCount === totalCount) {
+            rec.skipped(new SkipTestCase('all subcases were skipped'));
+          }
+        } else {
+          await this.runTest(
+            rec,
+            inst.forTest(),
+            /* throwSkip */ false,
+            getExpectedStatus(selfQuery)
+          );
         }
-        ++totalCount;
-      }
-      if (skipCount === totalCount) {
-        rec.skipped(new SkipTestCase('all subcases were skipped'));
+      } finally {
+        // Runs as long as the fixture constructor succeeded, even if initialization or the test failed.
+        await inst.doFinalize();
```

Comment:
I was thinking we would keep only the eventual expectations for a certain amount of cases in flight. But that may be difficult to do without extension changes to the fixture.